library/src/main/java/com/google/android/exoplayer/upstream/DefaultHttpDataSource.java

```diff
@@ -165,8 +166,22 @@ public long open(DataSpec dataSpec) throws HttpDataSourceException {
       throw new InvalidContentTypeException(contentType, dataSpec);
     }
 
+    // only used if gzip not used
+    long contentLength = getContentLength(connection);
+
+    if (responseCode == 200 && dataSpec.position != 0) {
```

Comment:
This doesn't consider range requests that start at offset 0 but impose an upper limit less than the length of the content. I think the condition should be:

if (responseCode == 200 && !(dataSpec.position == 0 && dataSpec.length != C.LENGTH_UNBOUNDED)) {
  ....
}

You then also need to impose the upper bound in the read method (that can happen unconditionally on all requests that impose such a limit).